src/python/pants/core/goals/test.py

```diff
@@ -204,6 +226,149 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
+                    subsystem: SuperTesterSubsystem,
+                ) -> Partitions[SuperTesterFieldSet]:
+                    if subsystem.skip:
+                        return Partitions()
+
+                    # One possible implementation
+                    return Partitions.partition_per_input(request.field_sets)
+
+        2. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestResult` instance. E.g.
+                @rule
+                async def super_test(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestResult:
+                    ...
+
+        3. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugRequest:
+                    ...
+
+        4. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugAdapterRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugAdapterRequest:
+                    ...
+
+    NOTE: If your tester doesn't support running in debug mode / running with the debug adapter, you
+        can have the `@rule`s in 3 and 4 raise `NotImplementedError`s.
+
+    Lastly, register the rules which tell Pants about your plugin.
+    E.g.
+        def rules():
+            return [
+                *collect_rules(),
+                *SuperTesterRequest.registration_rules(),
+            ]
+    """
+
+    name: ClassVar[str]
+    field_set_type: ClassVar[type[TestFieldSet]]
+
+    # Prevent this class from being detected by pytest as a test class.
+    __test__ = False
+
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class PartitionRequest(Generic[_FieldSetT]):
+        """Returns a unique `PartitionRequest` type per calling type.
+
+        This serves us 2 purposes:
+            1. `TestRequest.PartitionRequest` is the unique type used as a union base for plugin registration.
+            2. `<Plugin Defined Subclass>.PartitionRequest` is the unique type used as the union member.
+        """
+
+        field_sets: tuple[_FieldSetT, ...]
+
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class SubPartition(Generic[_FieldSetT]):
+        elements: Tuple[_FieldSetT, ...]
+        key: str
+
+    _PartitionRequestBase = PartitionRequest
+    _SubPartitionBase = SubPartition
+
+    if not TYPE_CHECKING:
+
+        @memoized_classproperty
+        def PartitionRequest(cls):
+            @union(in_scope_types=[EnvironmentName])
```

Comment:
I cargo-culted these `in_scope_types` settings from the `LintRequest` structure - I must admit I don't really know what it does or how it should interact with reintroducing environment-awareness to this new structure 😬

Reply:
These are necessary to run any `Process` inside the "subgraph", because `Process` requires the environment